File: client/src/main/java/io/xol/chunkstories/client/ingame/LocalPlayerImplementation.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.client.ingame;

import java.util.Iterator;

import io.xol.chunkstories.api.GameContext;
import io.xol.chunkstories.api.Location;
import io.xol.chunkstories.api.client.ClientInputsManager;
import io.xol.chunkstories.api.client.IngameClient;
import io.xol.chunkstories.api.client.LocalPlayer;
import io.xol.chunkstories.api.entity.Entity;
import io.xol.chunkstories.api.entity.traits.serializable.TraitControllable;
import io.xol.chunkstories.api.entity.traits.serializable.TraitInventory;
import io.xol.chunkstories.api.graphics.Window;
import io.xol.chunkstories.api.graphics.systems.dispatching.DecalsManager;
import io.xol.chunkstories.api.item.inventory.Inventory;
import io.xol.chunkstories.api.net.Packet;
import io.xol.chunkstories.api.particles.ParticlesManager;
import io.xol.chunkstories.api.sound.SoundManager;
import io.xol.chunkstories.api.world.World;
import io.xol.chunkstories.api.world.WorldClientNetworkedRemote;
import io.xol.chunkstories.api.world.WorldMaster;
import io.xol.chunkstories.client.LocalClientLoadingAgent;
import io.xol.chunkstories.client.ingame.IngameClientImplementation;
import io.xol.chunkstories.world.WorldClientCommon;

public class LocalPlayerImplementation implements LocalPlayer {
	final IngameClientImplementation client;
	final WorldClientCommon world;

	private Entity controlledEntity;

	public final LocalClientLoadingAgent loadingAgent;

	public LocalPlayerImplementation(IngameClientImplementation client, WorldClientCommon world) {
		this.client = client;
		this.world = world;

		this.loadingAgent = new LocalClientLoadingAgent(client, this, world);
	}

	@Override
	public ClientInputsManager getInputsManager() {
		return client.getInputsManager();
	}

	@Override
	public Entity getControlledEntity() {
		return controlledEntity;
	}

	@Override
	public boolean setControlledEntity(Entity entity) {
		TraitControllable ec = entity != null ? entity.traits.get(TraitControllable.class) : null;
		if (entity != null && ec != null) {
			this.subscribe(entity);

			// If a world master, directly set the entity's controller
			if (world instanceof WorldMaster)
				ec.setController(this);
			// In remote networked worlds, we need to subscribe the server to our player
			// actions to the controlled entity so he gets updates
			else if (entity.getWorld() instanceof WorldClientNetworkedRemote) {
				// When changing controlled entity, first unsubscribe the remote server from the
				// one we no longer own
				if (controlledEntity != null && entity != controlledEntity)
					((WorldClientNetworkedRemote) controlledEntity.getWorld()).getRemoteServer()
							.unsubscribe(controlledEntity);

				// Let know the server of new changes
				((WorldClientNetworkedRemote) entity.getWorld()).getRemoteServer().subscribe(entity);
			}

			controlledEntity = entity;
		} else if (entity == null && controlledEntity != null) {
			// Directly unset it
			if (world instanceof WorldMaster)
				controlledEntity.traits.with(TraitControllable.class, ec2 -> ec2.setController(null));
			// When loosing control over an entity, stop sending the server worldInfo about it
			else if (controlledEntity.getWorld() instanceof WorldClientNetworkedRemote)
				((WorldClientNetworkedRemote) controlledEntity.getWorld()).getRemoteServer()
						.unsubscribe(controlledEntity);

			controlledEntity = null;
		}

		return true;
	}

	@Override
	public SoundManager getSoundManager() {
		return client.getSoundManager();
	}

	@Override
	public ParticlesManager getParticlesManager() {
		return world.getParticlesManager();
	}

	@Override
	public DecalsManager getDecalsManager() {
		return world.getDecalsManager();
	}

	@Override
	public long getUUID() {
		return client.getUser().getName().hashCode();
	}

	@Override
	public Iterator<Entity> getSubscribedToList() {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public boolean subscribe(Entity entity) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	public boolean unsubscribe(Entity entity) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	public void unsubscribeAll() {
		// TODO Auto-generated method stub

	}

	@Override
	public void pushPacket(Packet packet) {
		// TODO Auto-generated method stub

	}

	public boolean isSubscribedTo(Entity entity) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	public boolean hasFocus() {
		return client.getGui().hasFocus();
	}

	@Override
	public String getName() {
		return client.getUser().getName();
	}

	@Override
	public String getDisplayName() {
		return getName();
	}

	@Override
	public void sendMessage(String msg) {
		client.print(msg);
	}

	@Override
	public Location getLocation() {
		Entity controlledEntity = this.controlledEntity;
		if (controlledEntity != null)
			return controlledEntity.getLocation();
		return null;
	}

	@Override
	public void setLocation(Location l) {
		Entity controlledEntity = this.controlledEntity;
		if (controlledEntity != null)
			controlledEntity.traitLocation.set(l);
	}

	@Override
	public boolean isConnected() {
		return true;
	}

	@Override
	public boolean hasSpawned() {
		return controlledEntity != null;
	}

	@Override
	public GameContext getContext() {
		return this.client;
	}

	@Override
	public World getWorld() {
		Entity controlledEntity = this.controlledEntity;
		if (controlledEntity != null)
			return controlledEntity.getWorld();
		return null;
	}

	@Override
	public boolean hasPermission(String permissionNode) {
		return true;
	}

	@Override
	public void flush() {
		// TODO Auto-generated method stub

	}

	@Override
	public void disconnect() {

	}

	@Override
	public void disconnect(String disconnectionReason) {

	}

	@Override
	public Window getWindow() {
		return client.getGameWindow();
	}

	@Override
	public void openInventory(Inventory inventory) {
		Entity entity = this.getControlledEntity();
		if (entity != null && inventory.isAccessibleTo(entity)) {
			// Directly open it without further concern
			// client.openInventories(inventory);

			TraitInventory TraitInventory = entity.traits.get(TraitInventory.class);

			if (TraitInventory != null)
				client.getGui().openInventories(TraitInventory, inventory);
			else
				client.getGui().openInventories(inventory);
		}
		// else
		// this.sendMessage("Notice: You don't have access to this inventory.");
	}

	@Override
	public IngameClient getClient() {
		return client;
	}
}

File: client/src/main/java/xyz/chunkstories/client/ingame/LocalPlayerImplementation.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.client.ingame

import xyz.chunkstories.PlayerCommon
import xyz.chunkstories.api.client.ClientInputsManager
import xyz.chunkstories.api.client.LocalPlayer
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.EntitySerialization
import xyz.chunkstories.api.entity.traits.serializable.TraitControllable
import xyz.chunkstories.api.entity.traits.serializable.TraitInventory
import xyz.chunkstories.api.graphics.Window
import xyz.chunkstories.api.item.inventory.Inventory
import xyz.chunkstories.api.net.Packet
import xyz.chunkstories.api.util.kotlin.toVec3f
import xyz.chunkstories.api.world.WorldClientNetworkedRemote
import xyz.chunkstories.api.world.WorldMaster
import xyz.chunkstories.sound.ALSoundManager
import xyz.chunkstories.world.WorldClientCommon

class LocalPlayerImplementation(override val client: IngameClientImplementation, internal val world: WorldClientCommon) : PlayerCommon(client.user.name), LocalPlayer {

    val loadingAgent: LocalClientLoadingAgent = LocalClientLoadingAgent(client, this, world)

    override val displayName: String
        get() = name
    override val inputsManager: ClientInputsManager
        get() = client.inputsManager
    override val subscribedToList: Collection<Entity>
        get() = TODO("not implemented") //To change initializer of created properties use File | Settings | File Templates.
    override val uuid: Long
        get() = client.user.name.hashCode().toLong()
    override val window: Window
        get() = client.gameWindow

    override var controlledEntity: Entity? = null
        set(new) {
            val old = field

            val ec = new?.traits?.get(TraitControllable::class.java)
            if (new != null && ec != null) {
                this.subscribe(new)

                // If a world master, directly set the entity's controller
                if (world is WorldMaster)
                    ec.controller = this
                else if (new.world is WorldClientNetworkedRemote) {
                    // When changing controlled entity, first unsubscribe the remote server from the
                    // one we no longer own
                    if (old != null && new !== old)
                        (old.world as WorldClientNetworkedRemote).remoteServer.unsubscribe(old)

                    // Let know the server of new changes
                    (new.world as WorldClientNetworkedRemote).remoteServer.subscribe(new)
                }// In remote networked worlds, we need to subscribe the server to our player
                // actions to the controlled entity so he gets updates

                field = new
            } else if (new == null && old != null) {
                // Directly unset it
                if (world is WorldMaster)
                    old.traits[TraitControllable::class]?.let { it.controller = null }
                else if (old.world is WorldClientNetworkedRemote)
                    (old.world as WorldClientNetworkedRemote).remoteServer
                            .unsubscribe(old)// When loosing control over an entity, stop sending the server updates about it

                field = null
            }

            return
        }

    init {
        eventEntersWorld(world)
    }

    fun update() {
        loadingAgent.updateUsedWorldBits()

        controlledEntity?.let { entity ->
            val camera = entity.traits[TraitControllable::class]?.camera
            camera?.let {
                (client.soundManager as ALSoundManager).setListenerPosition(camera.position.toVec3f(), camera.lookingAt, camera.up)
            }
        }
    }

    override fun subscribe(entity: Entity): Boolean {
        // TODO Auto-generated method stub
        return false
    }

    override fun unsubscribe(entity: Entity): Boolean {
        // TODO Auto-generated method stub
        return false
    }

    override fun unsubscribeAll() {
        // TODO Auto-generated method stub

    }

    override fun pushPacket(packet: Packet) {
        // TODO Auto-generated method stub

    }

    override fun hasFocus(): Boolean {
        return client.gui.hasFocus() && inputsManager.mouse.isGrabbed
    }

    override fun sendMessage(msg: String) {
        client.print(msg)
    }

    override fun hasPermission(permissionNode: String): Boolean {
        return true
    }

    override fun flush() {
        // TODO Auto-generated method stub

    }

    override fun openInventory(inventory: Inventory) {
        val entity = this.controlledEntity
        if (entity != null && inventory.isAccessibleTo(entity)) {
            val playerInventory = entity.traits[TraitInventory::class.java]?.inventory

            if (playerInventory != null)
                client.gui.openInventories(playerInventory, inventory)
            else
                client.gui.openInventories(inventory)
        }
    }

    fun destroy() {
        val playerWorldMetadata = world.playersMetadata[this]!!
        val controlledEntity1 = controlledEntity
        if (controlledEntity1 != null) {
            playerWorldMetadata.savedEntity = EntitySerialization.serializeEntity(controlledEntity1)
            this.controlledEntity = null
            //val playerEntityFile = File(world.folderPath + "/players/" + this.name.toLowerCase() + ".json")
            //EntityFileSerialization.writeEntityToDisk(playerEntityFile, playerEntity)
        }
        eventLeavesWorld(world)
        loadingAgent.unloadEverything(true)
    }
}
